@types/automerge/index.d.ts

```diff
@@ -204,76 +177,61 @@ declare module 'automerge' {
   }
 
   interface Change {
-    message?: string
-    requestType?: RequestType
+    message: string
     actor: string
+    time: number
     seq: number
-    deps: Clock
+    deps: Hash[]
     ops: Op[]
-    diffs?: Diff[]
   }
 
   interface Op {
     action: OpAction
-    obj: UUID
-    key?: string
-    value?: any
+    obj: OpId
+    key: string | number
+    insert: boolean
+    child?: OpId
+    value?: number | boolean | string | null
     datatype?: DataType
-    elem?: number
+    pred?: OpId[]
   }
 
   interface Patch {
     actor?: string
     seq?: number
-    clock?: Clock
-    deps?: Clock
-    canUndo?: boolean
-    canRedo?: boolean
-    diffs: Diff[]
+    clock: Clock
+    deps: Hash[]
+    diffs: ObjectDiff
   }
 
-  interface Diff {
-    action: DiffAction
+  interface ObjectDiff {
+    objectId: OpId
     type: CollectionType
-    obj: UUID
-    path?: string[]
-    key?: string
-    index?: number
-    value?: any
-    elemId?: string
-    conflicts?: Conflict[]
-    datatype?: DataType
-    link?: boolean
+    edits?: Edit[]
+    props?: {[propName: string]: {[opId: string]: ObjectDiff | ValueDiff}}
   }
 
-  interface Conflict {
-    actor: string
-    value: any
-    link?: boolean
+  interface ValueDiff {
+    value: number | boolean | string | null
+    datatype?: DataType
   }
 
-  type RequestType =
-    | 'change' //..
-    | 'redo'
-    | 'undo'
+  interface Edit {
+    action: 'insert' | 'remove'
+    index: number
+    elemId: OpId
```

Comment:
Is elemId really non optional or is there missing a `?`

At the following there is an edit created which does not includes an elmId.
https://github.com/automerge/automerge/blob/81079ff75d2234b47cb912bad728158f2e71c527/frontend/context.js#L424

Reply:
I believe it is optional for removes, the rust version may help https://github.com/automerge/automerge-rs/blob/c9ba656c0c7947c948b83a6d9432eb93e26c825b/automerge-protocol/src/lib.rs#L389-L398